File: apps/api-fastapi/wo_wf/transitions.py
"""Loop 7 Phase 7.2 — Python port of the WO/WF transition helpers.

Mirrors `packages/contracts/wo-wf/transitions.ts`. The declarative
state-machine tables live in `contracts.state_machines` and agree
byte-for-byte with the TS canonical source via the Phase 6.1 parity
test. The helper logic around the tables (load status, check
permissions, update status, optionally open an execution_cycle,
write audit) is Python-native.

Parity strategy (R-011 in CODEX log):
  - State-machine tables: authoritative Python tuples, parity-tested
    vs TS via `tests/contract/state-machine-parity.test.ts`.
  - Helper semantics: documented in ADR-017; both TS and Python
    helpers consume the same table and write audit rows with
    identical {machine, from, to, reason, cycleId?} metadata shape.
  - Structural drift protection: a future loop can add a fixture-
    driven transition parity test that fires both helpers against the
    same WO fixture + compares resulting row state.

Only the helpers Loop 7 routes need land in this first cut:
  - transition_work_order
  - watchdog_expire_work_order
  - transition_workflow
(transition_workflow_execution and transition_workflow_step_run can
land in Loop 9+ when the orchestrator calls them directly; Streamlit
Loop 8 does not drive those manually.)
"""

from __future__ import annotations

import json
from typing import Any, Optional

import asyncpg

from authz.audit_writer import write_audit_row
from authz.check_permission import UserGrant, check_permission_decide
from contracts.state_machines import STATE_MACHINES, TransitionSpec


class IllegalTransition(Exception):
    def __init__(
        self, *, machine: str, from_: str, to: str, code: str
    ) -> None:
        super().__init__(
            f"IllegalTransition on {machine}: {from_} → {to} ({code})"
        )
        self.machine = machine
        self.from_ = from_
        self.to = to
        self.code = code


class RowNotFound(Exception):
    def __init__(self, *, table: str, id: str, client_id: str) -> None:
        super().__init__(f"RowNotFound: {table} id={id} client_id={client_id}")
        self.table = table
        self.id = id
        self.client_id = client_id


class PermissionDenied(Exception):
    def __init__(
        self,
        *,
        user_id: str,
        client_id: str,
        permission: str,
        reason: str,
        role: Optional[str],
    ) -> None:
        super().__init__(
            f"PermissionDenied: user={user_id} client={client_id} "
            f"permission={permission} reason={reason} role={role}"
        )
        self.user_id = user_id
        self.client_id = client_id
        self.permission = permission
        self.reason = reason
        self.role = role


# ── Internals ────────────────────────────────────────────────────────────


async def _load_status(
    conn: asyncpg.Connection,
    table: str,
    row_id: str,
    client_id: Optional[str],
) -> str:
    if client_id is None:
        row = await conn.fetchrow(
            f"SELECT status::text AS status FROM {table} WHERE id = $1",
            row_id,
        )
    else:
        row = await conn.fetchrow(
            f"SELECT status::text AS status FROM {table} WHERE id = $1 AND client_id = $2",
            row_id,
            client_id,
        )
    if row is None:
        raise RowNotFound(
            table=table, id=row_id, client_id=client_id or "(nested)"
        )
    return row["status"]


def _first_spec(machine: str, from_state: str, to: str) -> TransitionSpec:
    if machine not in STATE_MACHINES:
        raise IllegalTransition(
            machine=machine, from_=from_state, to=to, code="unknown_machine"
        )
    if from_state == to:
        raise IllegalTransition(
            machine=machine, from_=from_state, to=to, code="illegal_transition"
        )
    saw_from = False
    for t in STATE_MACHINES[machine]:
        if t.from_ == from_state:
            saw_from = True
            if t.to == to:
                return t
    raise IllegalTransition(
        machine=machine,
        from_=from_state,
        to=to,
        code="illegal_transition" if saw_from else "unknown_from",
    )


async def _load_perm_context(
    conn: asyncpg.Connection,
    user_id: str,
    client_id: str,
    permission: str,
) -> dict:
    vocab = await conn.fetch("SELECT permission_key FROM permissions")
    known = [r["permission_key"] for r in vocab]
    mem_rows = await conn.fetch(
        """
        SELECT role::text AS role FROM client_memberships
        WHERE user_id = $1 AND client_id = $2 AND status = 'active'
        LIMIT 1
        """,
        user_id,
        client_id,
    )
    if not mem_rows:
        return {"role": None, "role_permissions": [], "user_grants": [], "known_permissions": known}
    role = mem_rows[0]["role"]
    role_rows = await conn.fetch(
        """
        SELECT p.permission_key
        FROM role_permissions rp
        JOIN permissions p ON p.id = rp.permission_id
        WHERE rp.role = $1::membership_role
        """,
        role,
    )
    grant_rows = await conn.fetch(
        """
        SELECT p.permission_key, pg.grant_type::text AS grant_type
        FROM permission_grants pg
        JOIN permissions p ON p.id = pg.permission_id
        WHERE pg.user_id = $1 AND pg.client_id = $2
        """,
        user_id,
        client_id,
    )
    return {
        "role": role,
        "role_permissions": [r["permission_key"] for r in role_rows],
        "user_grants": [
            UserGrant(
                permission_key=r["permission_key"], grant_type=r["grant_type"]
            )
            for r in grant_rows
        ],
        "known_permissions": known,
    }


async def _require_permissions(
    conn: asyncpg.Connection,
    *,
    user_id: str,
    client_id: str,
    permissions: tuple[str, ...],
    target_type: str,
    target_id: str,
    audit_metadata: dict[str, Any],
) -> None:
    for permission in permissions:
        ctx = await _load_perm_context(conn, user_id, client_id, permission)
        decision = check_permission_decide(
            role=ctx["role"],
            role_permissions=ctx["role_permissions"],
            user_grants=ctx["user_grants"],
            permission=permission,
            known_permissions=ctx["known_permissions"],
        )
        if decision.allowed:
            continue
        await write_audit_row(
            conn,
            client_id=client_id,
            actor_user_id=user_id,
            event="authz.denied",
            target_type=target_type,
            target_id=target_id,
            metadata={
                **audit_metadata,
                "permission": permission,
                "decision_reason": decision.reason,
                "role": decision.role,
            },
        )
        raise PermissionDenied(
            user_id=user_id,
            client_id=client_id,
            permission=permission,
            reason=decision.reason,
            role=decision.role,
        )


async def _insert_execution_cycle(
    conn: asyncpg.Connection,
    *,
    client_id: str,
    work_order_id: str,
    trigger: str,
    initiated_by_user_id: Optional[str],
    reason: Optional[str],
    metadata: dict[str, Any],
) -> str:
    prior = await conn.fetchrow(
        """
        SELECT id::text AS id, cycle_number
        FROM execution_cycles
        WHERE work_order_id = $1 AND client_id = $2
        ORDER BY cycle_number DESC
        LIMIT 1
        """,
        work_order_id,
        client_id,
    )
    next_cycle = (prior["cycle_number"] + 1) if prior else 1
    prior_id = prior["id"] if prior else None
    row = await conn.fetchrow(
        """
        INSERT INTO execution_cycles
          (client_id, work_order_id, cycle_number, trigger,
           initiated_by_user_id, reason, prior_cycle_id, metadata)
        VALUES ($1, $2, $3, $4::execution_cycle_trigger, $5, $6, $7, $8::jsonb)
        RETURNING id::text AS id
        """,
        client_id,
        work_order_id,
        next_cycle,
        trigger,
        initiated_by_user_id,
        reason,
        prior_id,
        json.dumps(metadata),
    )
    return row["id"]


# ── Public helpers ────────────────────────────────────────────────────────


async def transition_work_order(
    conn: asyncpg.Connection,
    *,
    work_order_id: str,
    client_id: str,
    actor_user_id: str,
    to: str,
    reason: Optional[str] = None,
) -> dict:
    from_status = await _load_status(
        conn, "work_orders", work_order_id, client_id
    )
    spec = _first_spec("work_order", from_status, to)
    audit_metadata: dict[str, Any] = {
        "machine": "work_order",
        "from": from_status,
        "to": to,
        "reason": reason,
    }
    await _require_permissions(
        conn,
        user_id=actor_user_id,
        client_id=client_id,
        permissions=spec.requires,
        target_type="work_order",
        target_id=work_order_id,
        audit_metadata=audit_metadata,
    )
    await conn.execute(
        """
        UPDATE work_orders
        SET status = $1, updated_at = now()
        WHERE id = $2 AND client_id = $3
        """,
        to,
        work_order_id,
        client_id,
    )
    cycle_id: Optional[str] = None
    if spec.requires_cycle:
        cycle_id = await _insert_execution_cycle(
            conn,
            client_id=client_id,
            work_order_id=work_order_id,
            trigger="reopen",
            initiated_by_user_id=actor_user_id,
            reason=reason,
            metadata={"from": from_status, "to": to, "event": spec.event},
        )
        audit_metadata["cycleId"] = cycle_id
    await write_audit_row(
        conn,
        client_id=client_id,
        actor_user_id=actor_user_id,
        event=spec.event,
        target_type="work_order",
        target_id=work_order_id,
        metadata=audit_metadata,
    )
    return {
        "from": from_status,
        "to": to,
        "event": spec.event,
        "cycle_id": cycle_id,
    }


async def watchdog_expire_work_order(
    conn: asyncpg.Connection,
    *,
    work_order_id: str,
    client_id: str,
    actor_user_id: str,
    reason: str,
) -> dict:
    from_status = await _load_status(
        conn, "work_orders", work_order_id, client_id
    )
    if from_status != "processing":
        raise IllegalTransition(
            machine="work_order",
            from_=from_status,
            to="blocked",
            code="illegal_transition",
        )
    spec = next(
        (
            t
            for t in STATE_MACHINES["work_order"]
            if t.from_ == "processing"
            and t.to == "blocked"
            and t.event == "work_order.watchdog_expired"
        ),
        None,
    )
    if spec is None:
        raise RuntimeError("watchdog spec missing from work_order table")

    audit_metadata: dict[str, Any] = {
        "machine": "work_order",
        "from": from_status,
        "to": "blocked",
        "reason": reason,
        "trigger": "watchdog",
    }
    await _require_permissions(
        conn,
        user_id=actor_user_id,
        client_id=client_id,
        permissions=spec.requires,
        target_type="work_order",
        target_id=work_order_id,
        audit_metadata=audit_metadata,
    )
    await conn.execute(
        """
        UPDATE work_orders
        SET status = 'blocked', updated_at = now()
        WHERE id = $1 AND client_id = $2
        """,
        work_order_id,
        client_id,
    )
    cycle_id = await _insert_execution_cycle(
        conn,
        client_id=client_id,
        work_order_id=work_order_id,
        trigger="watchdog",
        initiated_by_user_id=actor_user_id,
        reason=reason,
        metadata={"from": from_status, "to": "blocked", "event": spec.event},
    )
    audit_metadata["cycleId"] = cycle_id
    await write_audit_row(
        conn,
        client_id=client_id,
        actor_user_id=actor_user_id,
        event=spec.event,
        target_type="work_order",
        target_id=work_order_id,
        metadata=audit_metadata,
    )
    return {
        "from": from_status,
        "to": "blocked",
        "event": spec.event,
        "cycle_id": cycle_id,
    }


async def transition_workflow(
    conn: asyncpg.Connection,
    *,
    workflow_id: str,
    client_id: str,
    actor_user_id: str,
    to: str,
    reason: Optional[str] = None,
) -> dict:
    from_status = await _load_status(conn, "workflows", workflow_id, client_id)
    spec = _first_spec("workflow", from_status, to)
    audit_metadata: dict[str, Any] = {
        "machine": "workflow",
        "from": from_status,
        "to": to,
        "reason": reason,
    }
    await _require_permissions(
        conn,
        user_id=actor_user_id,
        client_id=client_id,
        permissions=spec.requires,
        target_type="workflow",
        target_id=workflow_id,
        audit_metadata=audit_metadata,
    )
    await conn.execute(
        """
        UPDATE workflows
        SET status = $1, updated_at = now()
        WHERE id = $2 AND client_id = $3
        """,
        to,
        workflow_id,
        client_id,
    )
    await write_audit_row(
        conn,
        client_id=client_id,
        actor_user_id=actor_user_id,
        event=spec.event,
        target_type="workflow",
        target_id=workflow_id,
        metadata=audit_metadata,
    )
    return {"from": from_status, "to": to, "event": spec.event}

File: packages/contracts/wo-wf/transitions.ts
/**
 * Loop 6 Phase 6.2 — lifecycle transition helpers.
 *
 * Runtime entry points for WO / WF / execution / step-run state
 * changes. Each helper:
 *
 *   1. Loads the current row + tenant-scopes the lookup by `client_id`.
 *   2. Calls `describeTransition(...)` against the Phase 6.1 tables.
 *      Illegal transitions throw `IllegalTransition` — nothing touches
 *      the DB in that branch.
 *   3. For each permission in `spec.requires` calls `requirePermission`
 *      — missing permissions throw `PermissionDenied` and write an
 *      `authz.denied` audit row (ADR-014). Every gate runs under the
 *      caller's transaction.
 *   4. Updates the row's `status` column (tenant-scoped WHERE).
 *   5. If `spec.requiresCycle`, opens a new `execution_cycles` row for
 *      WO-scoped machines (work_order only; workflow_execution's
 *      `requiresCycle` flag is currently documentation for Loop 9+
 *      orchestration — see R-009 in the CODEX log).
 *   6. Writes exactly one audit row using the transition's declared
 *      event name with `{machine, from, to, reason, cycleId?}` metadata.
 *
 * Helper semantics:
 *
 *   - `transitionWorkOrder` / `transitionWorkflow` /
 *     `transitionWorkflowExecution` / `transitionWorkflowStepRun` are
 *     the general-purpose state-change entry points. They pick the
 *     FIRST matching spec from the state-machine table, so
 *     processing → blocked resolves to the "regular block" spec with
 *     `work_order.blocked` event (NOT the watchdog one).
 *
 *   - `watchdogExpireWorkOrder` is a dedicated helper for the
 *     watchdog path (processing → blocked with
 *     `work_order.watchdog_expired` event + cycle row). Exposed
 *     separately so a future Loop-9 watchdog job calls it
 *     unambiguously; manual operator blocks continue to use
 *     `transitionWorkOrder`.
 *
 * Loop 6 intentionally does NOT ship a scheduled watchdog job —
 * Darrel's approved default (§Q1) says helper-only in Loop 6; the
 * scheduler lands in Loop 9.
 */

import type { PoolClient } from "pg";

import { AUDIT_EVENTS } from "../audit/events";
import { writeAuditRow } from "../audit/writer";
import { requirePermission } from "../authz/require_permission";
import {
  describeTransition,
  STATE_MACHINES,
  type StateMachineKey,
  type TransitionSpec,
} from "./state_machines";

// ── Error types ───────────────────────────────────────────────────────

export class IllegalTransition extends Error {
  readonly machine: StateMachineKey;
  readonly from: string;
  readonly to: string;
  readonly code: "unknown_from" | "illegal_transition";

  constructor(opts: {
    machine: StateMachineKey;
    from: string;
    to: string;
    code: "unknown_from" | "illegal_transition";
  }) {
    super(
      `IllegalTransition on ${opts.machine}: ${opts.from} → ${opts.to} (${opts.code})`
    );
    this.name = "IllegalTransition";
    this.machine = opts.machine;
    this.from = opts.from;
    this.to = opts.to;
    this.code = opts.code;
  }
}

export class RowNotFound extends Error {
  readonly table: string;
  readonly id: string;
  readonly clientId: string;

  constructor(opts: { table: string; id: string; clientId: string }) {
    super(
      `RowNotFound: ${opts.table} id=${opts.id} client_id=${opts.clientId}`
    );
    this.name = "RowNotFound";
    this.table = opts.table;
    this.id = opts.id;
    this.clientId = opts.clientId;
  }
}

// ── Shared transition plumbing ────────────────────────────────────────

interface TransitionContext {
  machine: StateMachineKey;
  /** Resolved transition spec from the Phase 6.1 table. */
  spec: TransitionSpec;
  /** from-status (current row value). */
  from: string;
  /** Target status (caller-provided). */
  to: string;
}

async function loadStatus(
  client: PoolClient,
  table: string,
  id: string,
  clientId: string
): Promise<string> {
  const { rows } = await client.query<{ status: string }>(
    `SELECT status FROM ${table} WHERE id = $1 AND client_id = $2`,
    [id, clientId]
  );
  if (rows.length === 0) {
    throw new RowNotFound({ table, id, clientId });
  }
  return rows[0].status;
}

/**
 * Nested tables (workflow_templates, workflow_step_runs) do not carry
 * `client_id` directly. Callers supply the client_id explicitly; this
 * helper looks up status via the table's own id column and trusts the
 * caller's tenant claim (which the audit row + RLS will double-check).
 */
async function loadStatusWithoutClientFilter(
  client: PoolClient,
  table: string,
  id: string
): Promise<string> {
  const { rows } = await client.query<{ status: string }>(
    `SELECT status FROM ${table} WHERE id = $1`,
    [id]
  );
  if (rows.length === 0) {
    throw new RowNotFound({ table, id, clientId: "(nested — not on table)" });
  }
  return rows[0].status;
}

function pickFirstSpec(
  machine: StateMachineKey,
  from: string,
  to: string
): TransitionSpec {
  const r = describeTransition(machine, from, to);
  if (!r.ok) {
    throw new IllegalTransition({ machine, from, to, code: r.reason });
  }
  return r.spec;
}

async function requireAllPermissions(
  client: PoolClient,
  input: {
    actorUserId: string;
    clientId: string;
    permissions: readonly string[];
    targetType: string;
    targetId: string;
    auditMetadata: Record<string, unknown>;
  }
): Promise<void> {
  for (const permission of input.permissions) {
    await requirePermission(client, {
      userId: input.actorUserId,
      clientId: input.clientId,
      permission,
      targetType: input.targetType,
      targetId: input.targetId,
      metadata: input.auditMetadata,
    });
  }
}

async function insertExecutionCycle(
  client: PoolClient,
  input: {
    clientId: string;
    workOrderId: string;
    trigger:
      | "new"
      | "retry"
      | "reopen"
      | "unblock"
      | "watchdog"
      | "admin_repair"
      | "candidate_request_more";
    initiatedByUserId: string | null;
    reason: string | null;
    metadata: Record<string, unknown>;
  }
): Promise<string> {
  // Compute next cycle_number + pick prior_cycle_id from the newest
  // existing cycle for this WO. Unique on (work_order_id, cycle_number)
  // prevents races — a concurrent insert would fail the uniq constraint
  // and the caller's transaction rolls back.
  const priorRes = await client.query<{
    id: string;
    cycle_number: number;
  }>(
    `SELECT id, cycle_number
     FROM execution_cycles
     WHERE work_order_id = $1 AND client_id = $2
     ORDER BY cycle_number DESC
     LIMIT 1`,
    [input.workOrderId, input.clientId]
  );
  const nextCycleNumber =
    priorRes.rows.length > 0 ? priorRes.rows[0].cycle_number + 1 : 1;
  const priorCycleId = priorRes.rows[0]?.id ?? null;

  const { rows } = await client.query<{ id: string }>(
    `INSERT INTO execution_cycles
       (client_id, work_order_id, cycle_number, trigger,
        initiated_by_user_id, reason, prior_cycle_id, metadata)
     VALUES ($1, $2, $3, $4::execution_cycle_trigger, $5, $6, $7, $8)
     RETURNING id`,
    [
      input.clientId,
      input.workOrderId,
      nextCycleNumber,
      input.trigger,
      input.initiatedByUserId,
      input.reason,
      priorCycleId,
      JSON.stringify(input.metadata),
    ]
  );
  return rows[0].id;
}

// ── Public helpers ────────────────────────────────────────────────────

export interface TransitionWorkOrderInput {
  workOrderId: string;
  clientId: string;
  actorUserId: string;
  to: string;
  reason?: string;
}

export interface TransitionResult {
  from: string;
  to: string;
  event: string;
  cycleId?: string;
}

export async function transitionWorkOrder(
  client: PoolClient,
  input: TransitionWorkOrderInput
): Promise<TransitionResult> {
  const from = await loadStatus(
    client,
    "work_orders",
    input.workOrderId,
    input.clientId
  );
  const spec = pickFirstSpec("work_order", from, input.to);

  const auditMetadata: Record<string, unknown> = {
    machine: "work_order",
    from,
    to: input.to,
    reason: input.reason ?? null,
  };

  await requireAllPermissions(client, {
    actorUserId: input.actorUserId,
    clientId: input.clientId,
    permissions: spec.requires,
    targetType: "work_order",
    targetId: input.workOrderId,
    auditMetadata,
  });

  await client.query(
    `UPDATE work_orders
     SET status = $1, updated_at = now()
     WHERE id = $2 AND client_id = $3`,
    [input.to, input.workOrderId, input.clientId]
  );

  let cycleId: string | undefined;
  if (spec.requiresCycle) {
    // First-match semantics mean this is the reopen path
    // (work_order.reopened). The watchdog path has its own helper.
    cycleId = await insertExecutionCycle(client, {
      clientId: input.clientId,
      workOrderId: input.workOrderId,
      trigger: "reopen",
      initiatedByUserId: input.actorUserId,
      reason: input.reason ?? null,
      metadata: { from, to: input.to, event: spec.event },
    });
    auditMetadata.cycleId = cycleId;
  }

  await writeAuditRow(client, {
    clientId: input.clientId,
    actorUserId: input.actorUserId,
    event: spec.event,
    targetType: "work_order",
    targetId: input.workOrderId,
    metadata: auditMetadata,
  });

  return { from, to: input.to, event: spec.event, cycleId };
}

export interface WatchdogExpireInput {
  workOrderId: string;
  clientId: string;
  /** agent_system actor id — used for audit + cycle lineage. */
  actorUserId: string;
  reason: string;
}

/**
 * Watchdog path — marks a processing WO blocked with the
 * `work_order.watchdog_expired` audit event + writes a cycle row
 * (trigger=watchdog) recording the timer breach. Callers are
 * responsible for firing this only against WOs they've decided are
 * stale; Loop 9+ orchestration adds the scheduled job that invokes it.
 */
export async function watchdogExpireWorkOrder(
  client: PoolClient,
  input: WatchdogExpireInput
): Promise<TransitionResult> {
  const from = await loadStatus(
    client,
    "work_orders",
    input.workOrderId,
    input.clientId
  );
  if (from !== "processing") {
    throw new IllegalTransition({
      machine: "work_order",
      from,
      to: "blocked",
      code: "illegal_transition",
    });
  }

  // Pick the watchdog-specific spec (not the first match).
  const spec = STATE_MACHINES.work_order.find(
    (t) =>
      t.from === "processing" &&
      t.to === "blocked" &&
      t.event === AUDIT_EVENTS.WORK_ORDER_WATCHDOG_EXPIRED
  );
  if (!spec) {
    // Defensive: state_machines table was tampered with between phases.
    throw new Error("watchdog transition spec missing from work_order table");
  }

  const auditMetadata: Record<string, unknown> = {
    machine: "work_order",
    from,
    to: "blocked",
    reason: input.reason,
    trigger: "watchdog",
  };

  await requireAllPermissions(client, {
    actorUserId: input.actorUserId,
    clientId: input.clientId,
    permissions: spec.requires,
    targetType: "work_order",
    targetId: input.workOrderId,
    auditMetadata,
  });

  await client.query(
    `UPDATE work_orders
     SET status = 'blocked', updated_at = now()
     WHERE id = $1 AND client_id = $2`,
    [input.workOrderId, input.clientId]
  );

  const cycleId = await insertExecutionCycle(client, {
    clientId: input.clientId,
    workOrderId: input.workOrderId,
    trigger: "watchdog",
    initiatedByUserId: input.actorUserId,
    reason: input.reason,
    metadata: { from, to: "blocked", event: spec.event },
  });
  auditMetadata.cycleId = cycleId;

  await writeAuditRow(client, {
    clientId: input.clientId,
    actorUserId: input.actorUserId,
    event: spec.event,
    targetType: "work_order",
    targetId: input.workOrderId,
    metadata: auditMetadata,
  });

  return { from, to: "blocked", event: spec.event, cycleId };
}

export interface TransitionWorkflowInput {
  workflowId: string;
  clientId: string;
  actorUserId: string;
  to: string;
  reason?: string;
}

export async function transitionWorkflow(
  client: PoolClient,
  input: TransitionWorkflowInput
): Promise<TransitionResult> {
  const from = await loadStatus(
    client,
    "workflows",
    input.workflowId,
    input.clientId
  );
  const spec = pickFirstSpec("workflow", from, input.to);

  const auditMetadata: Record<string, unknown> = {
    machine: "workflow",
    from,
    to: input.to,
    reason: input.reason ?? null,
  };

  await requireAllPermissions(client, {
    actorUserId: input.actorUserId,
    clientId: input.clientId,
    permissions: spec.requires,
    targetType: "workflow",
    targetId: input.workflowId,
    auditMetadata,
  });

  await client.query(
    `UPDATE workflows
     SET status = $1, updated_at = now()
     WHERE id = $2 AND client_id = $3`,
    [input.to, input.workflowId, input.clientId]
  );

  await writeAuditRow(client, {
    clientId: input.clientId,
    actorUserId: input.actorUserId,
    event: spec.event,
    targetType: "workflow",
    targetId: input.workflowId,
    metadata: auditMetadata,
  });

  return { from, to: input.to, event: spec.event };
}

export interface TransitionWorkflowExecutionInput {
  executionId: string;
  clientId: string;
  actorUserId: string;
  to: string;
  reason?: string;
}

export async function transitionWorkflowExecution(
  client: PoolClient,
  input: TransitionWorkflowExecutionInput
): Promise<TransitionResult> {
  const from = await loadStatus(
    client,
    "workflow_executions",
    input.executionId,
    input.clientId
  );
  const spec = pickFirstSpec("workflow_execution", from, input.to);

  const auditMetadata: Record<string, unknown> = {
    machine: "workflow_execution",
    from,
    to: input.to,
    reason: input.reason ?? null,
  };

  await requireAllPermissions(client, {
    actorUserId: input.actorUserId,
    clientId: input.clientId,
    permissions: spec.requires,
    targetType: "workflow_execution",
    targetId: input.executionId,
    auditMetadata,
  });

  await client.query(
    `UPDATE workflow_executions
     SET status = $1, updated_at = now()
     WHERE id = $2 AND client_id = $3`,
    [input.to, input.executionId, input.clientId]
  );

  // NOTE: `requiresCycle` on the failed → running spec is documentation
  // for a future Loop 9 orchestration helper; execution_cycles is
  // WO-scoped today so Phase 6.2 does not open a cycle row here.
  // See R-009 in IWO3_OVERNIGHT_DEV_RESOLUTION_LOG for rationale.

  await writeAuditRow(client, {
    clientId: input.clientId,
    actorUserId: input.actorUserId,
    event: spec.event,
    targetType: "workflow_execution",
    targetId: input.executionId,
    metadata: auditMetadata,
  });

  return { from, to: input.to, event: spec.event };
}

export interface TransitionWorkflowStepRunInput {
  stepRunId: string;
  clientId: string;
  actorUserId: string;
  to: string;
  reason?: string;
}

/**
 * workflow_step_runs does not carry a direct `client_id` column (the
 * tenant is derived via workflow_executions → client_id; see
 * migration 0006 RLS policy). The helper accepts `clientId` from the
 * caller for audit + authz purposes. RLS on `workflow_step_runs`
 * enforces the real tenant boundary on the UPDATE via the nested
 * EXISTS policy.
 */
export async function transitionWorkflowStepRun(
  client: PoolClient,
  input: TransitionWorkflowStepRunInput
): Promise<TransitionResult> {
  const from = await loadStatusWithoutClientFilter(
    client,
    "workflow_step_runs",
    input.stepRunId
  );
  const spec = pickFirstSpec("workflow_step_run", from, input.to);

  const auditMetadata: Record<string, unknown> = {
    machine: "workflow_step_run",
    from,
    to: input.to,
    reason: input.reason ?? null,
  };

  await requireAllPermissions(client, {
    actorUserId: input.actorUserId,
    clientId: input.clientId,
    permissions: spec.requires,
    targetType: "workflow_step_run",
    targetId: input.stepRunId,
    auditMetadata,
  });

  // lint:bypass-rls-explain="workflow_step_runs has no direct client_id column; tenant isolation is enforced via the RLS EXISTS policy against workflow_executions (migration 0006) + the caller's requirePermission gate above. The UPDATE is PK-scoped (id is globally unique) and RLS refuses to update a row whose parent workflow_execution.client_id does not match current_setting('app.current_client_id')."
  await client.query(
    `UPDATE workflow_step_runs
     SET status = $1, updated_at = now()
     WHERE id = $2`,
    [input.to, input.stepRunId]
  );

  await writeAuditRow(client, {
    clientId: input.clientId,
    actorUserId: input.actorUserId,
    event: spec.event,
    targetType: "workflow_step_run",
    targetId: input.stepRunId,
    metadata: auditMetadata,
  });

  return { from, to: input.to, event: spec.event };
}
